hbase/stackable/patches/2.6.1/0005-Allow-overriding-ipc-bind-port-and-use-alternative-p.patch

```diff
@@ -0,0 +1,195 @@
+     configurationManager.registerObserver(this.rpcServices);
+     configurationManager.registerObserver(this);
+@@ -586,8 +592,7 @@ public class HMaster extends HRegionServer implements MasterServices {
+       registerConfigurationObservers();
+       Threads.setDaemonThreadRunning(new Thread(() -> TraceUtil.trace(() -> {
+         try {
+-          int infoPort = putUpJettyServer();
+-          startActiveMasterManager(infoPort);
++          startActiveMasterManager(useThisInfoPortInstead);
+         } catch (Throwable t) {
+           // Make sure we log the exception.
+           String error = "Failed to become Active Master";
+@@ -2991,7 +2996,7 @@ public class HMaster extends HRegionServer implements MasterServices {
+         }
+         case MASTER_INFO_PORT: {
+           if (infoServer != null) {
+-            builder.setMasterInfoPort(infoServer.getPort());
++            builder.setMasterInfoPort(useThisInfoPortInstead);
+           }
+           break;
+         }
+diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java
+index 27bcef2f06..80949070fd 100644
+--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java
++++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java
+@@ -24,6 +24,7 @@ import static org.apache.hadoop.hbase.HConstants.DEFAULT_HBASE_SPLIT_WAL_MAX_SPL
+ import static org.apache.hadoop.hbase.HConstants.DEFAULT_SLOW_LOG_SYS_TABLE_CHORE_DURATION;
+ import static org.apache.hadoop.hbase.HConstants.HBASE_SPLIT_WAL_COORDINATED_BY_ZK;
+ import static org.apache.hadoop.hbase.HConstants.HBASE_SPLIT_WAL_MAX_SPLITTER;
++import static org.apache.hadoop.hbase.HConstants.REGIONSERVER_PORT;
+ import static org.apache.hadoop.hbase.master.waleventtracker.WALEventTrackerTableCreator.WAL_EVENT_TRACKER_ENABLED_DEFAULT;
+ import static org.apache.hadoop.hbase.master.waleventtracker.WALEventTrackerTableCreator.WAL_EVENT_TRACKER_ENABLED_KEY;
+ import static org.apache.hadoop.hbase.namequeues.NamedQueueServiceChore.NAMED_QUEUE_CHORE_DURATION_DEFAULT;
+@@ -505,6 +506,10 @@ public class HRegionServer extends Thread
+    */
+   protected String useThisHostnameInstead;
+ 
++  protected int useThisPortInstead;
++
++  protected int useThisInfoPortInstead;
++
+   /**
+    * @deprecated since 2.4.0 and will be removed in 4.0.0. Use
+    *             {@link HRegionServer#UNSAFE_RS_HOSTNAME_DISABLE_MASTER_REVERSEDNS_KEY} instead.
+@@ -669,6 +674,8 @@ public class HRegionServer extends Thread
+       this.namedQueueRecorder = NamedQueueRecorder.getInstance(this.conf);
+       rpcServices = createRpcServices();
+       useThisHostnameInstead = getUseThisHostnameInstead(conf);
++      useThisPortInstead = getUseThisPortInstead(conf);
++      useThisInfoPortInstead = conf.getInt("hbase.info.port" , this.infoServer != null ? this.infoServer.getPort() : -1);
```

Comment:
Do we want to pull `hbase.info.port` into a constant for consistency? For the same reason, we might want to split it into master/regionserver variants?

Also, directionally.. for RPC ports we follow the existing RPC _hostname_ convention of "leave the existing key as the advertised port, introduce a new key for overriding the bound port". I agree that overriding the advertised port instead makes more sense _in isolation_ (and is consistent with kafka et al), but we should probably try to be consistent (both with the rest of our patch and with existing HBase code).

Reply:
`hbase.info.port `is now `hbase.master.bound.info.port` and `hbase.regionserver.bound.info.port`.
We are not touching `MASTER_INFO_PORT` (= "hbase.master.info.port"): is that you meant with "leave the existing key as the advertised port, introduce a new key for overriding the bound port"?

Reply:
So
`public static final String MASTER_BOUND_INFO_PORT = "hbase.master.bound.info.port";`
should be renamed to something like:
`public static final String MASTER_ADVERTISED_INFO_PORT = "hbase.master.advertised.info.port";`
as `MASTER_INFO_PORT` is already the bound one?

Reply:
That would be one way, yeah. But as I said above, I think we should try to be consistent about whether the "old" property (`MASTER_INFO_PORT`) is for the bound or advertised port.

Reply:
I've endeavoured to make that change in a separate patch: ee7ea2e